File: chapters/18-event-driven-architecture/java/IdempotentEventConsumer.java
// Scenario: fulfillment handles OrderPlaced with at-least-once delivery from Kafka/SQS
// Demonstrates: idempotency via processed-event ledger before side effects
// Trade-off: extra storage vs duplicate shipments on redelivery

import java.sql.Connection;
import java.util.Optional;

public class IdempotentEventConsumer {

    public void onOrderPlaced(Connection conn, String eventId, String orderId) throws Exception {
        if (alreadyProcessed(conn, eventId)) {
            return;
        }
        conn.setAutoCommit(false);
        try {
            markProcessed(conn, eventId);
            try (var ps = conn.prepareStatement(
                    "INSERT INTO shipments (order_id, status) VALUES (?, 'PENDING')")) {
                ps.setString(1, orderId);
                ps.executeUpdate();
            }
            conn.commit();
        } catch (Exception e) {
            conn.rollback();
            throw e;
        }
    }

    private boolean alreadyProcessed(Connection conn, String eventId) throws Exception {
        try (var ps = conn.prepareStatement("SELECT 1 FROM processed_events WHERE event_id = ?")) {
            ps.setString(1, eventId);
            try (var rs = ps.executeQuery()) {
                return rs.next();
            }
        }
    }

    private void markProcessed(Connection conn, String eventId) throws Exception {
        try (var ps = conn.prepareStatement("INSERT INTO processed_events (event_id) VALUES (?)")) {
            ps.setString(1, eventId);
            ps.executeUpdate();
        }
    }
}

File: docs/chapters/18-event-driven-architecture/go/idempotent_event_consumer.go
// Scenario: fulfillment consumes OrderPlaced; broker may redeliver
// Demonstrates: check processed_events before creating shipment
// Trade-off: ledger table growth — partition/archive by TTL policy

package main

import (
	"context"
	"database/sql"
	"errors"
)

func onOrderPlaced(ctx context.Context, db *sql.DB, eventID, orderID string) error {
	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()

	var exists int
	err = tx.QueryRowContext(ctx,
		`SELECT 1 FROM processed_events WHERE event_id = $1`, eventID).Scan(&exists)
	if err == nil {
		return nil // duplicate delivery — safe no-op
	}
	if !errors.Is(err, sql.ErrNoRows) {
		return err
	}
	if _, err = tx.ExecContext(ctx,
		`INSERT INTO processed_events (event_id) VALUES ($1)`, eventID); err != nil {
		return err
	}
	if _, err = tx.ExecContext(ctx,
		`INSERT INTO shipments (order_id, status) VALUES ($1, 'PENDING')`, orderID); err != nil {
		return err
	}
	return tx.Commit()
}
